leaderelection/leader_election.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leaderelection
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"time"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog"
+)
+
+const (
+	defaultLeaseDuration = 15 * time.Second
+	defaultRenewDeadline = 10 * time.Second
+	defaultRetryPeriod   = 5 * time.Second
+)
+
+// leaderElection is a convenience wrapper around client-go's leader election library.
+type leaderElection struct {
+	runFunc func(ctx context.Context)
+
+	// the lockName identifies the leader election config and should be shared across all members
+	lockName string
+	// the identity is the unique identity of the currently running member
+	identity string
+	// the namespace to store the lock resource
+	namespace string
+	// resourceLock defines the type of leaderelection that should be used
+	// valid options are resourcelock.ConfigMapsResourceLock and resourcelock.EndpointsResourceLock
+	resourceLock string
+
+	leaseDuration time.Duration
+	renewDeadline time.Duration
+	retryPeriod   time.Duration
+
+	clientset kubernetes.Interface
+}
+
+func NewConfigMapLeaderElection(clientset kubernetes.Interface, lockName, lockNamespace string, runFunc func(ctx context.Context)) *leaderElection {
```

Comment:
I think we only want to support Lease going forward and not ConfigMap nor Endpoints

Reply:
I left this here for backwards compat because external attacher still uses ConfigMap and external provisioner still uses Endpoints. I updated the PR so that `NewLeaderElection` defaults to Lease type, I can remove it otherwise :)

Reply:
I'm not sure how much benefit we get from still having the option to support the other mechanisms. It means the sidecar rbacs will still be different and may behave differently (and have different bugs) depending on the mechanism.

Another interesting side-effect that I'm not sure how widespread it is, but if the CSI driver implements some sort of in-memory caching of state that is shared between provisioner/attach/resize/snapshot/etc, then the fact that different sidecars from different pods could be the active leader could cause inconsistencies. (ie provision called in one pod, attach called from another pod).  I guess for that case, we just need to say that if you support running your driver with leader election, then it needs to be resilient to this scenario.

Reply:
I'd like to use Lease everywhere and remove ConfigMap / Endpoints, if it is possible. Do we need to be backward compatible in this case? IMO the only change is RBAC.

Reply:
The risk is that you may have two leaders during a rolling update of the driver because some pods will use the old type, and the new ones will use Lease.

What is the consequence of having two leaders?  You may provision/delete/attach/detach twice?  I think provision/delete twice is safe to do, but I'm not sure about attach/detach twice.  Can all control planes handle it?

Reply:
PR updated to only support leases

Reply:
I'm still concerned that we could be breaking some driver for attach/detach that we don't know about.  Given that we never called out leader-election as an alpha/beta feature, I think we should play it on the safe side and deprecate the old way while introducing the option to use the new Lease.  Maybe it's worth having all the methods in the library then so at least the sidecar code is not super ugly.

Reply:
Sounds good to me, I'll update the PR!